core/bootstrap/src/main/java/org/eclipse/dataspaceconnector/system/runtime/BaseRuntime.java

```diff
@@ -0,0 +1,183 @@
+        runtime.boot();
+    }
+
+    protected Monitor getMonitor() {
+        return monitor;
+    }
+
+    /**
+     * Main entry point to runtime initialization. Calls all methods.
+     */
+    protected void boot() {
+        var typeManager = createTypeManager();
+        monitor = createMonitor();
+        MonitorProvider.setInstance(monitor);
+
+        var context = createContext(typeManager, monitor);
+        initializeContext(context);
+
+        var name = getRuntimeName(context);
+        try {
+            initializeVault(context);
+            List<ServiceExtension> serviceExtensions = createExtensions(context);
+            java.lang.Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdown(serviceExtensions, monitor)));
+            bootExtensions(context, serviceExtensions);
+
+        } catch (Exception e) {
+            onError(e);
+
+        }
+        monitor.info(format("%s ready", name));
+
+    }
+
+    /**
+     * Initializes the context. If {@link BaseRuntime#createContext(TypeManager, Monitor)} is overridden and the (custom) context
+     * needs to be initialized, this method should be overridden as well.
+     *
+     * @param context The context.
+     */
+    protected void initializeContext(ServiceExtensionContext context) {
+        if (context instanceof DefaultServiceExtensionContext) {
+            ((DefaultServiceExtensionContext) context).initialize();
+        }
+    }
+
+    /**
+     * The name of this runtime. This string is solely used for cosmetic/display/logging purposes.
+     * By default, {@link ServiceExtensionContext#getConnectorId()} is used.
+     */
+    protected String getRuntimeName(ServiceExtensionContext context) {
+        return context.getConnectorId();
+    }
+
+    /**
+     * Callback for any error that happened during runtime initialization
+     */
+    protected void onError(Exception e) {
+        monitor.severe("Error booting runtime", e);
+        System.exit(-1);  // stop the process
+    }
+
+    /**
+     * Starts all service extensions by invokint {@link org.eclipse.dataspaceconnector.system.ExtensionLoader#bootServiceExtensions(List, ServiceExtensionContext)}
+     *
+     * @param context           The {@code ServiceExtensionContext} that is used in this runtime.
+     * @param serviceExtensions a list of extensions
+     */
+    protected void bootExtensions(ServiceExtensionContext context, List<ServiceExtension> serviceExtensions) {
+        bootServiceExtensions(serviceExtensions, context);
+    }
+
+    /**
+     * Create a list of {@link ServiceExtension}s. By default this is done using the ServiceLoader mechanism. Override if
+     * e.g. a custom DI mechanism should be used.
+     *
+     * @param context A context to which all the service extensions should be registered
+     * @return a list of {@code ServiceExtension}s
+     */
+    @NotNull
+    protected List<ServiceExtension> createExtensions(ServiceExtensionContext context) {
```

Comment:
load or create?

Reply:
like before it is not necessary for an implementor to know that the default implementation actually _loads_ `ServiceExtensions`.

Reply:
You can also argue the other way round and state that it's an implementation detail if something needs to be created first or can be loaded from somewhere - at the end it's about loading.
But feels not important enough to discuss that :-) , I'm fine with it and will merge it as soon as you are ready.